design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,229 @@
+pc-q35-rhel9.2.0     RHEL-9.2.0 PC (Q35 + ICH9, 2009)
+pc-q35-rhel9.0.0     RHEL-9.0.0 PC (Q35 + ICH9, 2009)
+pc-q35-rhel8.6.0     RHEL-8.6.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.5.0     RHEL-8.5.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.4.0     RHEL-8.4.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.3.0     RHEL-8.3.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.2.0     RHEL-8.2.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.1.0     RHEL-8.1.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.0.0     RHEL-8.0.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel7.6.0     RHEL-7.6.0 PC (Q35 + ICH9, 2009) (deprecated)
+none                 empty machine
+```
+This means each workload must be updated to use a `pc-q35-rhel9.x.x` machine type.
+
+## Goals
+* Create an opt-in workflow that automates mass converting machine types of VMs
+	* if currently running, update VM machine type and add a label to alert the user the VM must be restarted for the change to take effect
+	* alternatively, if currently running, take VM offline immediately and update machine type
+	* if not currently running, update machine type immediately
+* Allow user to specify certain VMs for updating machine type automation (e.g. by namespace)
+
+## Non Goals
+While these additions could be beneficial to cluster-admins who wish to use the mass machine type transition (MMTT), these are not covered in the scope of this design proposal: 
+* Allowing user to filter VMs by running state: a cluster admin may want to only update VMs that are offline.
+* Allowing user to filter VMs in ways other than by namespace; e.g. with label-selector
+
+Also not covered in this design proposal is the method in which users/cluster admins will be informed that they have VMs with unsupported machine types.
+
+## User Stories
+As a cluster-admin, I want an automated way to update the machine type of all VMs to be compatible with CentOS Stream X without interrupting my workflow.
+As a cluster-admin, I want to be able to include only certain VMs to be updated (e.g. by namespace).
+As a cluster-admin, I want to be in control of the behavior of running VMs being updated; I can determine whether I want to manually restart running VMs being updated or let the automation restart all of them for me. 
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Create a new virtctl command that will allow the user to automatically update the machine type of any VMs with a machine type that is no longer supported. This command invokes a k8s `Job` that iterates through all VMs within a specified namespace (or all namespaces if none is specified), determines if the machine type is no longer supported, and updates it to the latest supported version. Both the minimum supported machine type and the target machine type are constants that can be updated as future versions are released and more machine types are deprecated upon these releases. 
```

Comment:
How does the system know what machine types are outdated on the given cluster?

And how does the system know to what version to update the machine type of a vm?
Would it be enough to just unset `machine.type` as our admission webhooks will auto fill it in with the most recent?

Please explain if the system has the knowledge (and how it is gaining it) or if the user is specifying this. And if we assuem the user, then where (the heck!) the user will get this info from.

Reply:
This was briefly discussed [above](https://github.com/kubevirt/community/pull/225#discussion_r1249356835). I completely agree that either a hard-coded version or a user-specified version (which, as you said, the user won't know where to get from) is a bad idea. 

As I see it, we have a few reasonable solutions:
1. Fetching the version dynamically. As @lyarwood [mentioned](https://github.com/kubevirt/community/pull/225#discussion_r1250506880), this is a bit tricky since we would need to support both downstream and upstream versions via regex or something similar. Saying that, if the version formats don't change too often and are pretty consistent this might not be such a bad idea.
2. Store the most-recent version in HCO on a per-release basis. This is also not great, as someone needs to update it every time a release takes place, which is error-prone.
3. Provide the user a list of possible versions, then asking him to specify the desired version. At least with this solution the admin would have a general idea which version to choose, and we can ease it even further with some documentation.

WDYT?

Reply:
I would go with approach 3 as IMO it is the most versatile and predictable one. You can set a filter to replace specific versions or a wildcard to replace all (which makes it behave like approach 1 again).

Reply:
At the moment, since KubeVirt knows the most recent machine type, the initial implementation can utilize that and also determine the minimum supported (i.e. oldest machine type that is not deprecated) from QEMU's list of supported machines to keep the initial implementation simple. However this is worth discussing and planning more and I would like to add it to the design as a follow up (as I have added in the Implementation Phases section of the design proposal).